src/backend/cuda/hamming.cu

```diff
@@ -0,0 +1,62 @@
+/*******************************************************
+ * Copyright (c) 2014, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#include <af/dim4.hpp>
+#include <af/defines.h>
+#include <ArrayInfo.hpp>
+#include <Array.hpp>
+#include <err_cuda.hpp>
+#include <handle.hpp>
+#include <kernel/hamming.hpp>
+#include <kernel/transpose.hpp>
+
+using af::dim4;
+
+namespace cuda
+{
+
+template<typename T>
+void hamming_matcher(Array<uint>& idx, Array<uint>& dist,
+                     const Array<T>& query, const Array<T>& train,
+                     const uint dist_dim, const uint n_dist)
+{
+    uint sample_dim = (dist_dim == 0) ? 1 : 0;
+    const dim4 qDims = query.dims();
+    const dim4 tDims = train.dims();
+
+    const dim4 queryTDims = dim4(qDims[1], qDims[0], qDims[2], qDims[3]);
+    const dim4 trainTDims = dim4(tDims[1], tDims[0], tDims[2], tDims[3]);
+    Array<T> queryT = createEmptyArray<T>(queryTDims);
+    Array<T> trainT = createEmptyArray<T>(trainTDims);
+    if (dist_dim == 0)
```

Comment:
This allocates memory on the device. You don't need it if dist_dim is not 0. Re-arrange the code to reduce the memory overhead.